newrelic/hooks/mlmodel_langchain.py

```diff
@@ -255,6 +259,476 @@ def wrap_similarity_search(wrapped, instance, args, kwargs):
+                    "tags": tags,
+                    "error": True,
+                }
+            )
+
+            transaction.record_custom_event("LlmChatCompletionSummary", full_chat_completion_summary_dict)
+
+            create_chat_completion_message_event(
+                transaction,
+                settings.app_name,
+                input_message_list,
+                completion_id,
+                span_id,
+                trace_id,
+                run_id,
+                conversation_id,
+                [],
+                message_ids,
+            )
+
+            raise
+
+    if not return_val:
+        return return_val
+
+    response = return_val
+    run_manager_info = getattr(transaction, "_nr_run_manager_info", {})
+    if hasattr(transaction, "_nr_run_manager_info"):
+        del transaction._nr_run_manager_info
+    run_id = run_manager_info.get("run_id", "")
+    metadata = run_manager_info.get("metadata", {})
+    tags = run_manager_info.get("tags", "")
+
+    input_message_list = [_input]
+    output_message_list = []
+    response_len = 1
+    try:
+        output_message_list = [response[0]] if response else []
+    except:
+        try:
+            output_message_list = [str(response)]
+        except:
+            _logger.warning("Unable to capture response. No response message event will be captured.")
+
+    # Make sure the builtin attributes take precedence over metadata attributes.
+    full_chat_completion_summary_dict = {"metadata.%s" % key: value for key, value in metadata.items()}
+    full_chat_completion_summary_dict.update(
+        {
+            "id": completion_id,
+            "appName": app_name,
+            "conversation_id": conversation_id,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction.guid,
+            "vendor": "langchain",
+            "ingest_source": "Python",
+            "virtual_llm": True,
+            "request_id": run_id,
+            "duration": ft.duration,
+            "response.number_of_messages": len(input_message_list) + len(output_message_list),
+            "tags": tags or "",
+        }
+    )
+
+    transaction.record_custom_event("LlmChatCompletionSummary", full_chat_completion_summary_dict)
+
+    create_chat_completion_message_event(
+        transaction,
+        settings.app_name,
+        input_message_list,
+        completion_id,
+        span_id,
+        trace_id,
+        run_id,
+        conversation_id,
+        output_message_list,
+        message_ids,
+    )
+
+    return return_val
+
+
+def wrap_chain_sync_run(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return wrapped(*args, **kwargs)
+
+    # Framework metric also used for entity tagging in the UI
+    transaction.add_ml_model_info("Langchain", LANGCHAIN_VERSION)
+
+    run_args = bind_args(wrapped, args, kwargs)
```

Comment:
I considered making a separate wrapper function but they were almost identical. I like the is_invoke idea-I think I'll take that approach. Definitely if these get more complicated in the future, we can separate them but for now I think they're so similar it's not worth it right now. Definitely re-using the functions make sense. I've actually been thinking about that when implementing sync vs async because basically the functions are identical except for the wrapped calls so I also like that approach. In general, I'd like to do some cleanup of all of our llm implementations before we release. There's a lot of ugliness in the defaulting None to "" that I'd like to get rid of. I'm thinking we should just push that change into our attribute handling. According to the agent spec, we shouldn't be converting None -> "None" or sending "" to the collector which is where a lot of this extra logic comes from right now. Those are both separate issues right now though but just some thoughts I have about it.